packages/contracts/src/contracts/current/protocol/Exchange/MixinExchangeCore.sol

```diff
@@ -77,134 +77,332 @@ contract MixinExchangeCore is
+            emit ExchangeStatus(uint8(Status.INVALID_TAKER), orderHash);
+            revert();
         }
-        require(takerAssetFillAmount > 0);
 
-        // Validate order expiration
-        if (block.timestamp >= order.expirationTimeSeconds) {
-            emit ExchangeError(uint8(Errors.ORDER_EXPIRED), orderHash);
-            return fillResults;
+        // Ensure valid takerAssetFillAmount
+        if(takerAssetFillAmount <= 0) {
+            emit ExchangeStatus(uint8(Status.TAKER_ASSET_FILL_AMOUNT_TOO_LOW), orderHash);
+            revert();
         }
+    }
 
-        // Validate order availability
-        uint256 remainingTakerAssetFillAmount = safeSub(order.takerAssetAmount, filled[orderHash]);
-        if (remainingTakerAssetFillAmount == 0) {
-            emit ExchangeError(uint8(Errors.ORDER_FULLY_FILLED), orderHash);
-            return fillResults;
+    /// @dev Calculates amounts filled and fees paid by maker and taker.
+    /// @param order to be filled.
+    /// @param orderStatus Status of order to be filled.
+    /// @param filledAmount Amount of order already filled.
+    /// @param takerAssetFillAmount Desired amount of order to fill by taker.
+    /// @return status Return status of calculating fill amounts. Returns Status.SUCCESS on success.
+    /// @return fillResults Amounts filled and fees paid by maker and taker.
+    function calculateFillAmounts(
+        Order memory order,
+        uint8 orderStatus,
+        uint256 filledAmount,
+        uint256 takerAssetFillAmount)
+        public
+        pure
+        returns (
+            uint8 status,
+            FillResults memory fillResults)
+    {
+        // Fill Amount must be greater than 0
+        if(takerAssetFillAmount <= 0) {
+            status = uint8(Status.TAKER_ASSET_FILL_AMOUNT_TOO_LOW);
+            return;
         }
 
+        // Ensure the order is fillable
+        if (orderStatus != uint8(Status.ORDER_FILLABLE)) {
+            status = uint8(orderStatus);
+            return;
+        }
+
+        // Compute takerAssetFilledAmount
+        uint256 remainingtakerAssetAmount = safeSub(order.takerAssetAmount, filledAmount);
+        fillResults.takerAssetFilledAmount = min256(takerAssetFillAmount, remainingtakerAssetAmount);
+
         // Validate fill order rounding
-        fillResults.takerAssetFilledAmount = min256(takerAssetFillAmount, remainingTakerAssetFillAmount);
-        if (isRoundingError(fillResults.takerAssetFilledAmount, order.takerAssetAmount, order.makerAssetAmount)) {
-            emit ExchangeError(uint8(Errors.ROUNDING_ERROR_TOO_LARGE), orderHash);
-            fillResults.takerAssetFilledAmount = 0;
-            return fillResults;
+        if (isRoundingError(
+            fillResults.takerAssetFilledAmount,
+            order.takerAssetAmount,
+            order.makerAssetAmount))
+        {
+            status = uint8(Status.ROUNDING_ERROR_TOO_LARGE);
```

Comment:
We need to reset `fillResults.takerAssetFilledAmount` to 0 before returning here.